vfio/src/vfio.rs

```diff
@@ -0,0 +1,596 @@
+        }
+    }
+
+    fn get_device(&self, name: &Path) -> Result<File> {
+        let uuid_osstr = name.file_name().ok_or(VfioError::InvalidPath)?;
+        let uuid_str = uuid_osstr.to_str().ok_or(VfioError::InvalidPath)?;
+        let path: CString = CString::new(uuid_str.as_bytes()).expect("CString::new() failed");
+        let path_ptr = path.as_ptr();
+
+        // Safe as we are the owner of self and path_ptr which are valid value.
+        let ret = unsafe { ioctl_with_ptr(self, VFIO_GROUP_GET_DEVICE_FD(), path_ptr) };
+        if ret < 0 {
+            return Err(VfioError::GroupGetDeviceFD);
+        }
+
+        // Safe as ret is valid FD
+        Ok(unsafe { File::from_raw_fd(ret) })
+    }
+}
+
+impl AsRawFd for VfioGroup {
+    fn as_raw_fd(&self) -> RawFd {
+        self.group.as_raw_fd()
+    }
+}
+
+impl Drop for VfioGroup {
+    fn drop(&mut self) {
+        self.kvm_device_del_group();
+        self.unset_container();
+    }
+}
+
+struct VfioRegion {
+    flags: u32,
+    size: u64,
+    offset: u64,
+}
+
+/// Vfio device for exposing regions which could be read/write to kernel vfio device.
+pub struct VfioDevice {
+    device: File,
+    group: VfioGroup,
+    regions: Vec<VfioRegion>,
+    mem: GuestMemoryMmap,
+}
+
+impl VfioDevice {
+    /// Create a new vfio device, then guest read/write on this device could be
+    /// transfered into kernel vfio.
+    /// sysfspath specify the vfio device path in sys file system.
+    pub fn new(sysfspath: &Path, vm: &VmFd, mem: GuestMemoryMmap) -> Result<Self> {
+        let mut uuid_path = PathBuf::new();
+        uuid_path.push(sysfspath);
+        uuid_path.push("iommu_group");
+        let group_path = uuid_path.read_link().map_err(|_| VfioError::InvalidPath)?;
+        let group_osstr = group_path.file_name().ok_or(VfioError::InvalidPath)?;
+        let group_str = group_osstr.to_str().ok_or(VfioError::InvalidPath)?;
+        let group_id = group_str
+            .parse::<u32>()
+            .map_err(|_| VfioError::InvalidPath)?;
+
+        let group = VfioGroup::new(group_id, vm)?;
+        let device = group.get_device(sysfspath)?;
+        let regions = Self::get_regions(&device)?;
+
+        Ok(VfioDevice {
+            device,
+            group,
+            regions,
+            mem,
+        })
+    }
+
+    /* Only support MSI with eventfd
+    fn get_irq_info(&self, count :&mut u32) -> Result<u32, VfioError>{
+        let mut irq_info = vfio_irq_info {
+                argsz: mem::size_of::<vfio_irq_info>() as u32,
+                flags: 0,
+                index: VFIO_PCI_MSI_IRQ_INDEX,
+                count: 0,
+        };
+        let ret = unsafe {
+            ioctl_with_mut_ref(self, VFIO_DEVICE_GET_IRQ_INFO(), &mut irq_info) };
+        if ret < 0 || irq_info.flags & VFIO_IRQ_INFO_EVENTFD == 0 {
+            return Err(VfioError::VfioDeviceGetIrqInfo)
+        }
+
+        *count = irq_info.count;
+
+        return Ok(irq_info.count)
+    } */
```

Comment:
Since this is commented out as not working, we should remove it.

Reply:
fixed.